From 534eb41b5bb0b86382a262e7410700016d62a834 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Tue, 21 Apr 2009 15:50:51 +0200
Subject: [PATCH] OOXML: eradicate SheetViewSettings::maSheetDatas vector.  Fixes n#494603.

Keeping gratituous rData = maSheetData references and gratituous
blocks (eg ::importPane) to minimize maSheetDatas -> mxSheetData diff.
When this passes all tests etc., feel free to remove, rename and
reindent.
---
 oox/inc/oox/xls/viewsettings.hxx |    6 +---
 oox/source/xls/viewsettings.cxx  |   63 ++++++++++++-------------------------
 2 files changed, 22 insertions(+), 47 deletions(-)

diff --git oox/inc/oox/xls/viewsettings.hxx oox/inc/oox/xls/viewsettings.hxx
index 8b9c604..750866b 100644
--- oox/inc/oox/xls/viewsettings.hxx
+++ oox/inc/oox/xls/viewsettings.hxx
@@ -146,11 +146,7 @@ public:
     void                finalizeImport();
 
 private:
-    OoxSheetViewDataRef createSheetViewData();
-
-private:
-    typedef RefVector< OoxSheetViewData > SheetViewDataVec;
-    SheetViewDataVec    maSheetDatas;
+    OoxSheetViewDataRef mxSheetData;
 };
 
 // ============================================================================
diff --git oox/source/xls/viewsettings.cxx oox/source/xls/viewsettings.cxx
index e19fa0d..3d56af7 100644
--- oox/source/xls/viewsettings.cxx
+++ oox/source/xls/viewsettings.cxx
@@ -270,14 +270,18 @@ OoxSheetSelectionData& OoxSheetViewData::createSelectionData( sal_Int32 nPaneId
 
 // ----------------------------------------------------------------------------
 
-SheetViewSettings::SheetViewSettings( const WorksheetHelper& rHelper ) :
-    WorksheetHelper( rHelper )
+SheetViewSettings::SheetViewSettings( const WorksheetHelper& rHelper )
+    : WorksheetHelper( rHelper )
+    , mxSheetData( new OoxSheetViewData )
 {
 }
 
 void SheetViewSettings::importSheetView( const AttributeList& rAttribs )
 {
-    OoxSheetViewData& rData = *createSheetViewData();
+    /* Use rData and keep gratituous blocks (eg ::importPane) to
+       minimize maSheetDatas -> mxSheetData diff.  When this passes
+       all tests etc., feel free to rename and reindent.  */
+    OoxSheetViewData& rData = *mxSheetData;
     rData.maGridColor.setIndexed( rAttribs.getInteger( XML_colorId, OOX_COLOR_WINDOWTEXT ) );
     rData.maFirstPos        = getAddressConverter().createValidCellAddress( rAttribs.getString( XML_topLeftCell, OUString() ), getSheetIndex(), false );
     rData.mnWorkbookViewId  = rAttribs.getToken( XML_workbookViewId, 0 );
@@ -298,16 +302,14 @@ void SheetViewSettings::importSheetView( const AttributeList& rAttribs )
 
 void SheetViewSettings::importTabColor( const AttributeList& rAttribs )
 {
-    OoxSheetViewData& rData = maSheetDatas.empty() ? *createSheetViewData() : *maSheetDatas.back();
+    OoxSheetViewData& rData = *mxSheetData;
     rData.maTabColor.importColor( rAttribs );
 }
 
 void SheetViewSettings::importPane( const AttributeList& rAttribs )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importPane - missing view data" );
-    if( !maSheetDatas.empty() )
     {
-        OoxSheetViewData& rData = *maSheetDatas.back();
+        OoxSheetViewData& rData = *mxSheetData;
         rData.maSecondPos    = getAddressConverter().createValidCellAddress( rAttribs.getString( XML_topLeftCell, OUString() ), getSheetIndex(), false );
         rData.mnActivePaneId = rAttribs.getToken( XML_activePane, XML_topLeft );
         rData.mnPaneState    = rAttribs.getToken( XML_state, XML_split );
@@ -318,12 +320,10 @@ void SheetViewSettings::importPane( const AttributeList& rAttribs )
 
 void SheetViewSettings::importSelection( const AttributeList& rAttribs )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importSelection - missing view data" );
-    if( !maSheetDatas.empty() )
     {
         // pane this selection belongs to
         sal_Int32 nPaneId = rAttribs.getToken( XML_pane, XML_topLeft );
-        OoxSheetSelectionData& rSelData = maSheetDatas.back()->createSelectionData( nPaneId );
+        OoxSheetSelectionData& rSelData = mxSheetData->createSelectionData( nPaneId );
         // cursor position
         rSelData.maActiveCell = getAddressConverter().createValidCellAddress( rAttribs.getString( XML_activeCell, OUString() ), getSheetIndex(), false );
         rSelData.mnActiveCellId = rAttribs.getInteger( XML_activeCellId, 0 );
@@ -335,7 +335,7 @@ void SheetViewSettings::importSelection( const AttributeList& rAttribs )
 
 void SheetViewSettings::importChartSheetView( const AttributeList& rAttribs )
 {
-    OoxSheetViewData& rData = *createSheetViewData();
+    OoxSheetViewData& rData = *mxSheetData;
     rData.mnWorkbookViewId = rAttribs.getToken( XML_workbookViewId, 0 );
     rData.mnCurrentZoom    = rAttribs.getInteger( XML_zoomScale, 100 );
     rData.mbSelected       = rAttribs.getBool( XML_tabSelected, false );
@@ -344,7 +344,7 @@ void SheetViewSettings::importChartSheetView( const AttributeList& rAttribs )
 
 void SheetViewSettings::importSheetView( RecordInputStream& rStrm )
 {
-    OoxSheetViewData& rData = *createSheetViewData();
+    OoxSheetViewData& rData = *mxSheetData;
     sal_uInt16 nFlags;
     sal_Int32 nViewType;
     BinAddress aFirstPos;
@@ -371,16 +371,14 @@ void SheetViewSettings::importSheetView( RecordInputStream& rStrm )
 
 void SheetViewSettings::importTabColor( RecordInputStream& rStrm )
 {
-    OoxSheetViewData& rData = maSheetDatas.empty() ? *createSheetViewData() : *maSheetDatas.back();
+    OoxSheetViewData& rData = *mxSheetData;
     rData.maTabColor.importColor( rStrm );
 }
 
 void SheetViewSettings::importPane( RecordInputStream& rStrm )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importPane - missing view data" );
-    if( !maSheetDatas.empty() )
     {
-        OoxSheetViewData& rData = *maSheetDatas.back();
+        OoxSheetViewData& rData = *mxSheetData;
 
         BinAddress aSecondPos;
         sal_Int32 nActivePaneId;
@@ -395,12 +393,10 @@ void SheetViewSettings::importPane( RecordInputStream& rStrm )
 
 void SheetViewSettings::importSelection( RecordInputStream& rStrm )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importSelection - missing view data" );
-    if( !maSheetDatas.empty() )
     {
         // pane this selection belongs to
         sal_Int32 nPaneId = rStrm.readInt32();
-        OoxSheetSelectionData& rSelData = maSheetDatas.back()->createSelectionData( lclGetOoxPaneId( nPaneId, -1 ) );
+        OoxSheetSelectionData& rSelData = mxSheetData->createSelectionData( lclGetOoxPaneId( nPaneId, -1 ) );
         // cursor position
         BinAddress aActiveCell;
         rStrm >> aActiveCell >> rSelData.mnActiveCellId;
@@ -415,7 +411,7 @@ void SheetViewSettings::importSelection( RecordInputStream& rStrm )
 
 void SheetViewSettings::importChartSheetView( RecordInputStream& rStrm )
 {
-    OoxSheetViewData& rData = *createSheetViewData();
+    OoxSheetViewData& rData = *mxSheetData;
     sal_uInt16 nFlags;
     rStrm >> nFlags >> rData.mnCurrentZoom >> rData.mnWorkbookViewId;
 
@@ -425,8 +421,7 @@ void SheetViewSettings::importChartSheetView( RecordInputStream& rStrm )
 
 void SheetViewSettings::importWindow2( BiffInputStream& rStrm )
 {
-    OSL_ENSURE( maSheetDatas.empty(), "SheetViewSettings::importWindow2 - multiple WINDOW2 records" );
-    OoxSheetViewData& rData = *createSheetViewData();
+    OoxSheetViewData& rData = *mxSheetData;
     if( getBiff() == BIFF2 )
     {
         rData.mbShowFormulas = rStrm.readuInt8() != 0;
@@ -481,15 +476,13 @@ void SheetViewSettings::importWindow2( BiffInputStream& rStrm )
 
 void SheetViewSettings::importPane( BiffInputStream& rStrm )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importPane - missing leading WINDOW2 record" );
-    if( !maSheetDatas.empty() )
     {
         sal_uInt8 nActivePaneId;
         sal_uInt16 nSplitX, nSplitY;
         BinAddress aSecondPos;
         rStrm >> nSplitX >> nSplitY >> aSecondPos >> nActivePaneId;
 
-        OoxSheetViewData& rData = *maSheetDatas.back();
+        OoxSheetViewData& rData = *mxSheetData;
         rData.mfSplitX = nSplitX;
         rData.mfSplitY = nSplitY;
         rData.maSecondPos = getAddressConverter().createValidCellAddress( aSecondPos, getSheetIndex(), false );
@@ -499,25 +492,21 @@ void SheetViewSettings::importPane( BiffInputStream& rStrm )
 
 void SheetViewSettings::importScl( BiffInputStream& rStrm )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importScl - missing leading WINDOW2 record" );
-    if( !maSheetDatas.empty() )
     {
         sal_uInt16 nNum, nDenom;
         rStrm >> nNum >> nDenom;
         OSL_ENSURE( nDenom > 0, "SheetViewSettings::importScl - invalid denominator" );
         if( nDenom > 0 )
-            maSheetDatas.back()->mnCurrentZoom = getLimitedValue< sal_Int32, sal_uInt16 >( (nNum * 100) / nDenom, 10, 400 );
+            mxSheetData->mnCurrentZoom = getLimitedValue< sal_Int32, sal_uInt16 >( (nNum * 100) / nDenom, 10, 400 );
     }
 }
 
 void SheetViewSettings::importSelection( BiffInputStream& rStrm )
 {
-    OSL_ENSURE( !maSheetDatas.empty(), "SheetViewSettings::importPane - missing leading WINDOW2 record" );
-    if( !maSheetDatas.empty() )
     {
         // pane this selection belongs to
         sal_uInt8 nPaneId = rStrm.readuInt8();
-        OoxSheetSelectionData& rSelData = maSheetDatas.back()->createSelectionData( lclGetOoxPaneId( nPaneId, -1 ) );
+        OoxSheetSelectionData& rSelData = mxSheetData->createSelectionData( lclGetOoxPaneId( nPaneId, -1 ) );
         // cursor position
         BinAddress aActiveCell;
         sal_uInt16 nActiveCellId;
@@ -534,8 +523,7 @@ void SheetViewSettings::importSelection( BiffInputStream& rStrm )
 
 void SheetViewSettings::finalizeImport()
 {
-    // force creation of sheet view data to get the Excel defaults
-    OoxSheetViewDataRef xData = maSheetDatas.empty() ? createSheetViewData() : maSheetDatas.front();
+    OoxSheetViewDataRef xData = mxSheetData;
 
     // #i59590# #158194# special handling for chart sheets (Excel ignores some settings in chart sheets)
     if( getSheetType() == SHEETTYPE_CHARTSHEET )
@@ -658,15 +646,6 @@ void SheetViewSettings::finalizeImport()
     getViewSettings().setSheetViewSettings( getSheetIndex(), xData, Any( aSheetProps.createPropertySequence() ) );
 }
 
-// private --------------------------------------------------------------------
-
-OoxSheetViewDataRef SheetViewSettings::createSheetViewData()
-{
-    OoxSheetViewDataRef xData( new OoxSheetViewData );
-    maSheetDatas.push_back( xData );
-    return xData;
-}
-
 // ============================================================================
 
 OoxWorkbookViewData::OoxWorkbookViewData() :
-- 
1.6.0.rc1.49.g98a8

